rclcpp/test/rclcpp/test_node.cpp

```diff
@@ -2455,3 +2490,89 @@ TEST_F(TestNode, get_publishers_subscriptions_info_by_topic) {
     subscription_list = node->get_subscriptions_info_by_topic("13");
   }, rclcpp::exceptions::InvalidTopicNameError);
 }
+
+TEST_F(TestNode, callback_groups) {
+  auto node = std::make_shared<rclcpp::Node>("node", "ns");
+  size_t num_callback_groups_in_basic_node = node->get_callback_groups().size();
+
+  auto group1 = node->create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive);
+  EXPECT_EQ(1u + num_callback_groups_in_basic_node, node->get_callback_groups().size());
+
+  auto group2 = node->create_callback_group(rclcpp::CallbackGroupType::Reentrant);
+  EXPECT_EQ(2u + num_callback_groups_in_basic_node, node->get_callback_groups().size());
+}
+
+// This is tested more thoroughly in node_interfaces/test_node_graph
+TEST_F(TestNode, get_entity_names) {
+  auto node = std::make_shared<rclcpp::Node>("node", "ns");
+  const auto node_names = node->get_node_names();
+  EXPECT_NE(
+    node_names.end(),
+    std::find(node_names.begin(), node_names.end(), node->get_fully_qualified_name()));
+
+  const auto topic_names_and_types = node->get_topic_names_and_types();
+  EXPECT_EQ(topic_names_and_types.end(), topic_names_and_types.find("topic"));
+
+  EXPECT_EQ(0u, node->count_publishers("topic"));
+  EXPECT_EQ(0u, node->count_subscribers("topic"));
+
+  const auto service_names_and_types = node->get_service_names_and_types();
+  EXPECT_EQ(service_names_and_types.end(), service_names_and_types.find("service"));
+
+  const auto service_names_and_types_by_node =
+    node->get_service_names_and_types_by_node("node", "/ns");
+  EXPECT_EQ(
+    service_names_and_types_by_node.end(),
+    service_names_and_types_by_node.find("service"));
+}
+
+TEST_F(TestNode, wait_for_graph_event) {
+  // Even though this node is only used in the std::thread below, it's here to ensure there is no
+  // race condition in its destruction and modification of the node_graph
+  auto node = std::make_shared<rclcpp::Node>("node", "ns");
+
+  constexpr std::chrono::seconds timeout(10);
+  auto thread_start = std::chrono::steady_clock::now();
+  auto thread_completion = thread_start;
+
+  // This runs until the graph is updated
+  std::thread graph_event_wait_thread([&thread_completion, node, timeout]() {
+      auto event = node->get_graph_event();
+      EXPECT_NO_THROW(node->wait_for_graph_change(event, timeout));
+      thread_completion = std::chrono::steady_clock::now();
+    });
+
+  // Start creating nodes until event triggers in graph_event_wait_thread or until 100 nodes have
+  // been created
+  std::vector<std::shared_ptr<rclcpp::Node>> nodes;
+  while (thread_completion == thread_start && nodes.size() < 100) {
+    nodes.emplace_back(std::make_shared<rclcpp::Node>("node"_unq, "ns"));
+    std::this_thread::sleep_for(std::chrono::milliseconds(1));
```

Comment:
I'm concerned that this will be a flaky test, particularly on Windows.  Can we really create 100 nodes in under 10 seconds reliably?  Out of curiosity, why did you choose 100, and not something like 10?

Reply:
This test isn't about creating 100 nodes, in fact on my machine only one was created before the test succeeded. The race condition is that `node->wait_for_graph_change()` may be called after the first node is created. So this while loop continues to create nodes until `wait_for_graph_change()` detects a change in the node graph. 100 was chosen as a sufficiently large number to assert the failure case. 10 would also work, but with a higher number, this test will be theoretically less flaky.

Reply:
Ah, that makes a lot more sense.  Thanks for the explanation.  The comment right above didn't make that totally clear to me; maybe update it to say something like:

```
  // Start creating nodes until at least one event triggers in graph_event_wait_thread or until 100 nodes have
  // been created (at which point this is a failure)
```

That being said, that isn't blocking, so I'll still approve.

Reply:
Updated comment with this language